### DISCUSSION TITLE
#GiveUpGitHub

### DISCUSSION THREAD

**Discussion by renerocksai** [nvim-telekasten/telekasten.nvim#149] [Announcements]
Hi telekasten enthusiasts,

I have not been involved a lot recently, but thanks to @lambtho12
development of telekasten.nvim has continued. Huge kudos to @lambtho12.
I encourage @lambtho12 to set up a buymeacoffee account and everyone who
can afford it to send him a little thankyou! 😊 That's not to say that
he's been the only contibutor - so, actually, kudos to all of you! 👏
Now, let's go and buy each other some coffees 😊.

I am in a very, very busy phase at the moment, so I have almost no time
left for contributing to telekasten. This is likely to change around
August / September 2022 when my current research project has "gone
live". At least, it's supposed to do that by the end of August.

Moving away from GitHub

Recent developments around GitHub and its Copilot have lead to the
`#GiveUpGitHub` campaign. Read more about it [here](https://giveupgithub.org).

TL;DR: I get it, I support it, I want to get away from GH. The question
is: where to go? After evaluating, I see 2 options:
[Codeberg.org](https://codeberg.org) and
[sourcehut.org](https://sourcehut.org). While I prefer sourcehut, I
think many would prefer Codeberg because of its familiar web interface.

Here's my take on the whole situation: the only thing that speaks for
GitHub is also one of its weak points: it has become the go-to place for
open source code repositories. With features like stars and web-based
pull-requests, code-review, etc. it is luring us to go to the website as
often as possible. I don't particularly like that. I also don't like
that they train their AI with our code and then go off and make money
with it, not giving a dime about free software license obligations.

I am not a huge fan of GitHub-style pull-requests (git itself supports
'native' pull-requests) and git was not intended to be used that way
anyway. So I can happily live without that.

The more I looked into sourcehut, the more I started to love its fresh
take on everything. I find its e-mail and patch based workflows superior
to anything GH, GitLab, Codeberg, etc have to offer. It is simpler and
easier, more open (no account required), and more robust: it's really
hard to delete a distributed, decentralized mailing list - unless we all
have microsoft office 365 accounts. 😊

Discussions on GitHub are just awful. There is no threading view. It
becomes a total mess when trying to reply to past messages that are not
the most recent one. **E-Mail has solved this problem half a century
ago**.

Just imagine: there's a mailing list where all discussions are hosted
and you can use it with your favourite e-mail client. Same goes for
issues: web and mail interface. If it's a big or chatty project, there
can be lists like announce, devel, discuss, tracker.

For me, it would also mean: not having to context-switch from terminal
(code, mail) to the browser for every issue, pull-request, etc.

So, to cut a long story short: I want telekasten.nvim to find a new
home. My number one choice would be [sourcehut](https://sourcehut.org).
Option number two would be [codeberg.org](https://codeberg.org).

To me, Codeberg would be: just moving, no improvements. Sourcehut would
mean: moving, with huge improvements. But: initially more unfamiliar.

For a fair assessment: it seems to be possible to import a GitHub
project to Codeberg, so that issues, for instance, get migrated, too.
However, I would not shy away to migrate issues to sourcehut manually.

What do you think?

So you get a feeling for what it's like, I have mirrored the
telekasten.nvim project to sourcehut [here](https://sr.ht/~renerocksai/telekasten.nvim/).

Please tell me your thoughts on moving away from GitHub using the
[telekasten.nvim-discuss mailing list](https://lists.sr.ht/~renerocksai/telekasten.nvim-discuss). Also
please note that sourcehut enforces text-only e-mail, so please make
sure you don't send HTML e-mail there.

I don't want a simple poll. I prefer arguments.

Thanks for reading that far and looking forward to your ideas!
-Rene

**Comment by renerocksai** [comment]
Just posted a reply considering user and contributer perspective, [here](https://lists.sr.ht/~renerocksai/telekasten.nvim-discuss/%3C20220709001858.zy5at6pgnzvea7px%40renerocks.ai%3E#%3C20220709112944.tnstr7hivr7dvhs3@renerocks.ai%3E).

Please feel fee to send your thoughts there, too. I'd be really interested in how you think about moving away from GH in general and the options we have.

**Comment by lambtho12** [comment]
Hi @renerocksai!

Thank you very much for your kind words! I did what I could with the little time I have. Hopefully I can manage to get some more time in two-three months as well. I don't drink that much coffee though, so if people like my work and want to make a small gesture, I would be delighted if they can send a few coins to their charity of choice instead!

------
Now, regarding the core concern here, **I COULD NOT BE HAPPIER with leaving Github**. Fuck them, their telemetry and the shameless exploitation of FOSS contributors and in general fuck Microsoft and their shitty, bloated ~~spywares~~ products. I only maintain this account to collaborate on other people's project (mostly this one). I have always despised them and agree with all the points made in the website you linked. I have personally used Gitlab for years now. Even though they are not the ideal solution, I still feel they are much more aligned with my values than Github. As a FOSS enthusiast and privacy advocate in general, I will gladly shift to sourcehut for this project (and who knows, maybe for the rest of my repo if all goes well)!

Regarding the practical details and considerations for this transition, I will discuss that over the mailing list so we can directly move away from these awful "discussions". :smile:

**Reply by renerocksai** [reply]
You speak from my heart! :heart: 

And please, by no means feel pressured to keep up the pace on contributions. In my eyes, telekasten does all I need it to do for my current needs. It will improve, I will participate more in the future - but no need to rush things. This is our project and we are merely humans with jobs and lives.

Regarding coffees: your body, your choice! :joy: But you got the gist of it. Animating people to give to good causes is a good cause in itself! So: my respects!

I am so glad to hear you're on the same/similar page wrt GH. Can't wait to read your thoughts!

**Comment by imbolc** [comment]
"I have not been involved a lot recently" .. and now I want to make contributions harder and nullify discoverability of the project for ideological reasons. Please, don't :)

**Reply by lambtho12** [reply]
Fair point.

I would argue that there is not that much contributors anyway. Without counting Rene, only 2 people have committed more than 100 lines (3 if we count the pending PRs from Ernie). I don't really think it will annoy a lot of people... Simple users will only have to change one line in their plugin managers and it will be all good for them. As discussed in the mailing list, the current idea is that this repo will continue to exist on GH so it will not break users configs (but it will no longer be updated).

Regarding discoverability, I personally have absolutely zero clue if people really learn about TK by typing "zettelkasten" in Github search bar.
For what it is worth, I just tested it: searching for `zettelkasten` on GH will output telekasten as the last result of page 5. `zettelkasten +vim` does not list it and only `zettelkasten + nvim` list it in the first page (first place). From basic intuition, I think most users come from places such as reddit, zettelkasten.de or regular search engines. I personally have _never_ used GH search to find for a nvim plugin.

Finally, why should anyone dismiss their values (or _ideologies_ as you call it), especially when they offer their work for free to anyone? AFAIK, the project is under MIT. You can fork it and re-host it on GH if you really want.

**Reply by imbolc** [reply]
> I would argue that there is not that much contributors anyway, only 2 people have committed more than 100 lines

I'm not sure locs is a fare metric here, even a single line (e.g. fixing a security issue) could be a big contribution, maybe even a review with no changes. Also don't forget about opened issues. But let's assume it's so, still, is it a valid reason to make contributions harder?

> Regarding discoverability

It's not only (or mostly at least for me) how to learn about a project, but rather on how to choose one among competitors. And I don't see analogous to github metrics here.

> `zettelkasten +vim`

Maybe it worth to add `vim` as a tag

> why should anyone dismiss their values

I'd expect people to periodically review their values. But in this case I'm rather pointing at a possible conflict of values.

> or ideologies as you call it

I'd say ideology is a way of expressing values. I'd consider even FOSS being an ideology, let alone a particular code hosting as an expression of this ideology. So even assuming values to be a completely personal choice, ideologies and their expressions still could be a matter of a rational discussion (based on how they fit into the values).

Coming back, I'd argue that **discoverability and collaboration are probably among core values behind FOSS**. Why would one even make their code public other way?

**Reply by renerocksai** [reply]
> "I have not been involved a lot recently" .. and now I want to make contributions harder and nullify the discoverability of the project for ideological reasons. Please, don't :)

Well, it's my project. Feel free to fork it and do better. 

As to discoverability: if you think long enough: maybe that's an issue that should be addressed elsewhere. I hardly use github to find stuff. I use a search engine which then points me to GitHub.

I don't think that the discoverability will be nullified. Especially since we can still leave a README pointing to SH on GH. At least for a while.

It isn't my intention to become famous with Telekasten or telekasten itself to become famous. I put it on GitHub with the best intentions that others might find it useful. And I think that worked out OK.

I would prefer to discuss on the advertised mailing list. Threading here is garbage.

**Reply by renerocksai** [reply]
> I would argue that there is not that much contributors anyway. Without counting Rene, only 2 people have committed more than 100 lines (3 if we count the pending PRs from Ernie). I don't really think it will annoy a lot of people... Simple users will only have to change one line in their plugin managers and it will be all good for them. As discussed in the mailing list, the current idea is that this repo will continue to exist on GH so it will not break users configs (but it will no longer be updated).

Reasonable to me.

> Regarding discoverability, I personally have absolutely zero clue if people really learn about TK by typing "zettelkasten" in Github search bar. For what it is worth, I just tested it: searching for `zettelkasten` on GH will output telekasten as the last result of page 5. `zettelkasten +vim` does not list it and only `zettelkasten + nvim` list it in the first page (first place). From basic intuition, I think most users come from places such as reddit, zettelkasten.de or regular search engines. I personally have _never_ used GH search to find for a nvim plugin.

100% ack.

> Finally, why should anyone dismiss their values (or _ideologies_ as you call it), especially when they offer their work for free to anyone? AFAIK, the project is under MIT. You can fork it and re-host it on GH if you really want.

More on ideologies later.

**Reply by renerocksai** [reply]
> > I would argue that there is not that much contributors anyway, only 2 people have committed more than 100 lines
> 
> I'm not sure locs is a fare metric here, even a single line (e.g. fixing a security issue) could be a big contribution, maybe even a review with no changes. Also don't forget about opened issues. But let's assume it's so, still, is it a valid reason to make contributions harder?

Let's talk about telekasten here and not some fictional repo pulled out of thin air: as far as I know, no security issues have been fixed by one-liners. IMHO, sloc are a fair measure. Not an ideal one, but a fair one to assess code contributions. Which we as people who write the code seem to value. There are other valuable contributions, I acknowledge that. Especially issues can be very useful for later reference. 

As explained on the mailing list, I can migrate all existing issues and PRs over to SH. So nothing valuable gets lost. Apart from discussions. Should we find that the discussions are so valuable that they would need to be migrated, too, then I think this could be done. It didn't seem worth it given the very limited time I and most of the code contributors have.

> Maybe it worth to add `vim` as a tag

No. Vim doesn't support lua, telescope, any of this. It would be misleading. I have added neovim, nvim though.

ideology vs values: ideologies can become dangerous, intrinsic values rather not. But that's totally not the point here.

Consider this: I found a place that seems to be a better home for my project. Now I want to move. But I still like to know other people's opinions since the new home and what comes with it is also new to me. In addition, many things that I found problematic with GH have now reached a new level. I stayed on GH when it became MS, but I wouldn't want to stay now - especially since now there are much better alternatives (real alternatives, not clones) out there.

> ... rational discussions ...

Yes please. Let's have them. On the mailinglist since this interface sucks. And also: let's leave ideologies out of it.

> Coming back, I'd argue that **discoverability and collaboration are probably among core values behind FOSS**. Why would one even make their code public other way?

Let's not argue over the ranking of perceived foss values. Let's get back to the project at hand: telekasten.

I made my plugin open for FOSS reasons, yes. They made sense: put it out there and if it's useful, you might even get interesting and useful stuff back. In the process, it might become useful for even more people - and that's a good thing. It turned out very positively. 

Now, if I had more time, I would try to keep every momentum going, but, at the moment, I just can't.

That said, everyone is free to "fork" and start their own momentum. I'd probably support it if some contributor came along and showed how they'd do all the stuff we don't have too much time for at the moment. Of course, there will be times where I and maybe others will have more time again. But having it here out in the open at least gives people a chance to use it and modify it to their likings - or fix the issues they have with it.

Anyway, this was never meant to be a social project with the goal to attract as many people as possible and involve them in discussions or anything else. It started as my idea for a cool plugin and it grew because of FOSS and GH.

If we move to SH, only the people who think SH is completely useless for them will not be able to make the jump. If one seriously thinks that writing an e-mail is harder than clicking around a web interface using sub-standard web editors, they cannot be helped. Using SH's issue interface is not hard. Raising issues there is easy. Contributing code is not hard. For code contributors, I've written about how to do that on the mailing list and even put some stuff in a [wiki](https://man.sr.ht/~renerocksai/migrate-to-sourcehut/). 

The fear that we move away and nobody will follow - I don't see it. Especially not in the long run. I expect SH to become more prominent once it's out of beta. Even now with the #GiveUpGitHub campaing, quite some seem to have made the jump.

Soon we'll have Codebergs, Sourehuts, GitLabs, GitHubs, etc for free projects - and we'll learn to not only rely on the GitHub search feature when searching for FOSS.

**Reply by renerocksai** [reply]
Two more things I forgot to address:

Code hosting as ideology: No, it's not. I, for my part, want nothing to do with GH exploiting the work of open source projects and making money with it, not giving a shit about license obligations. Thinking about it, it's ridiculous to "demand" from open source projects that they should stay on GitHub, the very place where the exploitation of open source software is taken to the next level.

Stars: Drew Devault calls them dopamine dispensors - and he is right. While they are a neat feature, they can incentivize project owners, hooked on dopamine, to rather attract stars instead of focusing on the more important stuff. 

I get it: if you like a project, you like it to do well: get all the stars it might deserve and all the discoverability and social buzz in the world. 

If GitHub-discoverability and GitHub-stars come with the cost of supporting everything else GitHub does, then, in my strong opinion, the price is too high. 

Having learned about SourceHut, I really like it there. I like the workflows, the thinking that went into everything, etc. It is more open than GitHub, as it encourages you to contribute via e-mail if you have no and want no account. It has no starring feature by choice - it's not a social platform. Let's earn the stars elsewhere. :smile:

**Reply by imbolc** [reply]
> I hardly use github to find stuff. I use a search engine which then points me to GitHub.

You haven't addressed the main point though. Let's say you search engine gave you 100 projects. Github gives me a simple positive-false filter via starts, forks etc. How are you going to evaluate these 100 projects without it?

> I would prefer to discuss on the advertised mailing list. Threading here is garbage.

See, that's an issue already. Maybe I'm ok with the threading and don't like mailing lists. Or maybe I like some third options. It's like formating wars we had before the era of opinionated formatters.

> If we move to SH, only the people who think SH is completely useless for them will not be able to make the jump.

Or people who think Github is good enough and don't want to spend any time to form an opinion on SH, don't forget about opportunity cost.

> Soon we'll have Codebergs, Sourehuts, GitLabs, GitHubs, etc for free projects - and we'll learn to not only rely on the GitHub search feature when searching for FOSS.

I hope this won't happen, I really don't want to lose Github's almost universal rating system, nor I want learning 10 different interfaces of small, buggy code hostings.

There's always people who want to benefit from splitting a community on a first error made by a big guy. Let's see how they behave when they grow big enough. I'm not saying stay with Github no matter what, but I think the bigger company become and the longer it lasts, the more mistakes it makes. Any company. So I'm comparing them by the benefits / mistakes ratio. And for me Github is much higher by this ratio.

**Reply by renerocksai** [reply]
> You haven't addressed the main point though. Let's say you search engine gave you 100 projects. Github gives me a simple positive-false filter via starts, forks etc. How are you going to evaluate these 100 projects without it?

I don't. I never had that problem. I am not a collector of open source projects either that I can star to show my support.

> > I would prefer to discuss on the advertised mailing list. Threading here is garbage.
> 
> See, that's an issue already. Maybe I'm ok with the threading and don't like mailing lists. Or maybe I like some third options. It's like formating wars we had before the era of opinionated formatters.

See, that is your opinion. Fair enough. It is one of many opinions to have. The difference is: that this is my project, so it's my decision to make. And if I can't stand GH discussions, your opinion and the way you have expressed it, is not wanting me to stay either. This has nothing to do with formating wars. Not having a threaded view is far worse than some silly formating. For discussions, where we refer to posts of different authors at different times, this interface here is just the worst. If you had at least brought up a better example. But GH discussions I wouldn't miss one bit.

> > If we move to SH, only the people who think SH is completely useless for them will not be able to make the jump.
> 
> Or people who think Github is good enough and don't want to spend any time to form an opinion on SH, don't forget about opportunity cost.

Their loss. I can't fix society, nor only wanting to stay with what one knows because anything else seems too much of a hassle. Honestly, I don't care too much. Anyone who uses telekasten is clearly capable of clicking on a link to a mailing list or to a different bug tracker. Those who want to stay ignorant: fair enough. Stay, don't learn, don't broaden your horizon. 

> > Soon we'll have Codebergs, Sourehuts, GitLabs, GitHubs, etc for free projects - and we'll learn to not only rely on the GitHub search feature when searching for FOSS.
> 
> I hope this won't happen, I really don't want to lose Github's almost universal rating system, nor I want learning 10 different interfaces of small, buggy code hostings.

Describing GH stars (at least I think you mean them) as universal rating system not worth losing disqualifies you of a sober discussion here. I am sorry. You realize there's a whole open source world out there that cannot be found on GitHub? Think GitLab for a moment. 

The ignorance with which you describe alternatives you seemingly self-proclaimingly haven't even checked out as small and buggy code hostings is another reason for not taking you seriously. You know you throw GitLab, Gitea, Codeberg, and all the others into the same boat with your statement? I hope that with time you'll grow (up) in that regard.

The solution is easy: ignore them! All of them! Never use real search-engines, and make GitHub your homepage and go-to place for everything.

> There's always people who want to benefit from splitting a community on a first error made by a big guy. Let's see how they behave when they grow big enough. I'm not saying stay with Github no matter what, but I think the bigger company become and the longer it lasts, the more mistakes it makes. Any company. So I'm comparing them by the benefits / mistakes ratio. And for me Github is much higher by this ratio.

You haven't thought about the broader issues here. You think it's about some big guy who made a mistake? Or maybe a second mistake? And we should give them some leeway? Are you serious?

It's also not about splitting a community. I am not in a GitHub community! Do you really think we're all in a GitHub community?

Man, this is my code. If I don't want it on GitHub, I leave. Git is flexible to cope with that. Git users might be too. Telekasten will leave GitHub - and so far the best option I can see is SourceHut. "But I love stars", "but I use GH to search" is not a convincing argument, neither is "GH discussions are great, they have formating". Decentralization can be less convenient than a 1-stop-open-source-shop. But, as most things, this can be fixed outside of GH, outside of all code hosting platforms.

Please stick with GitHub, but please also accept that I don't need more of your opinion on the matter at hand (which is: where to move? Staying on GH is not an option; thanks for your opinion anyway). Think aboud the decentralized nature of git, and where "pull requests" come from if you like. Centralizing "all" code at GitHub was never a good idea in the first place. 

Also, feel free to fork and host on GH. I will not stop you.

**Comment by lambtho12** [comment]
As per https://github.com/renerocksai/telekasten.nvim/discussions/196#discussioncomment-4601002, we decided to stop pursuing the #GiveUpGithup campaign. 

Even though we still feel strongly against Github and their actions, it was decided to keep telekasten here in order to avoid bothering the ever-growing user base too much.

Links to the campaign will be removed shortly from the Readme and docs to avoid any confusion.

**Comment by aniketgm** [comment]
So you guys gave up on #GiveUpGithub. LoL the irony !! 😄😄 
I too don't have any affiliation for Github, and I can understand why you guys decided not to switch.
I'm on Github only because of the good collaborators like you guys and many others who are still on Github. Many of the software's and plugins I use are on Github (sadly). However, my personal projects are all on SH and GL. Only the ones I want to make public are here.

If all or most of the collaborators Or even just the ones that contribute to the software's I use, if they move to other platform. Then I'm happily switching there with them. Till then GH it is.
